gms/feature_service.hh

```diff
@@ -144,6 +144,12 @@ public:
     gms::feature fragmented_commitlog_entries { *this, "FRAGMENTED_COMMITLOG_ENTRIES"sv };
     gms::feature maintenance_tenant { *this, "MAINTENANCE_TENANT"sv };
 
+    // Enable rack-aware view updates pairing for tablets
+    // when using NetworkTopologyStrategy (always with tablets at this time)
+    // and when the replication factor is a multiple of the number of racks
+    // in the datacenter.
+    gms::feature tablets_rack_aware_view_pairing { *this, "TABLETS_RACK_AWARE_VIEW_PAIRING"sv };
```

Comment:
> Having the rack aware view pairing as a cluster feature will change the pairing after upgrade. I'm wondering if this could cause inconsistencies in the view when an update is performed while some nodes have already enabled this feature and some didn't.
> 
> I might be missing some insight on how cluster features are getting enabled, aside from the fact that they wait until the entire cluster is upgraded and that they're propagated by gossip. Is there some mechanism which guarantees that all replicas performing a write have the same status of the feature?

Not as far as I know.

> If so there's probably no issue and you can ignore what I wrote below. Otherwise, it looks like we may in fact lose some view replica writes.

view update are lossy by nature as are updates to the base, and repair of the base the the view table is supposed to bring the view to consistency with the base and consistent with itself (across the view replicas).

> 
> For example, aview replica will miss the update if we perform a base write while one replica has this feature enabled and the other hasn't and as a result of the pairing change they swap their paired view replicas.
> 
> In all cases this can be fixed by performing repair on the views (maybe we already do that after upgrade?).

I don't think we usually recommend that but we could for this specific upgrade path.
Ideally, in the future, incremental repair should run frequently enough to fix the inconsistencies in a timely manner.

> Alternatively, during upgrade, we could also send updates to both paired view replicas - from the original pairing and the rack-aware pairing.

Nice idea.  Not sure if it's worth it...
We could also "multi-cast" to all replicas in the datacenter during the upgrade.

> 
> In almost all cases the update will also land on most of the view replicas - I can only think of one extreme scenario where it lands on exactly half (if RF is even, half of the nodes have enabled the feature, and all nodes in these halves switch pairings to the other half), so QUORUM reads are probably safe.
> 
> Alternatively, we could use this only for new tables - though that would be far less beneficial. And if we allowed to enable this on per-table basis we would face the same issues, so cluster feature is probably still the best approach, but we need to make sure the issue of losing writes is handled.

Since this is for tablets, and they are still not GA'ed in enterprise releases, practically it IS enabled for new tables.

Reply:
> > For example, aview replica will miss the update if we perform a base write while one replica has this feature enabled and the other hasn't and as a result of the pairing change they swap their paired view replicas.
> > In all cases this can be fixed by performing repair on the views (maybe we already do that after upgrade?).
> 
> I don't think we usually recommend that but we could for this specific upgrade path. Ideally, in the future, incremental repair should run frequently enough to fix the inconsistencies in a timely manner.

If we don't add the recommendation, I think it's at least worth mentioning somewhere (maybe commit message/cover letter) so that we can keep track of this if we try to improve the base-view consistency in the future

> > Alternatively, during upgrade, we could also send updates to both paired view replicas - from the original pairing and the rack-aware pairing.
> 
> Nice idea. Not sure if it's worth it...
We could also "multi-cast" to all replicas in the datacenter during the upgrade.
>
> > Alternatively, we could use this only for new tables - though that would be far less beneficial. And if we allowed to enable this on per-table basis we would face the same issues, so cluster feature is probably still the best approach, but we need to make sure the issue of losing writes is handled.
> 
> Since this is for tablets, and they are still not GA'ed in enterprise releases, practically it IS enabled for new tables.

Ok, then it's probably fine to leave this as is